source/common/router/config_impl.h

```diff
@@ -236,16 +236,18 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
 
   struct VirtualClusterBase : public VirtualCluster {
   public:
-    VirtualClusterBase(Stats::StatName stat_name, Stats::ScopePtr&& scope,
-                       const VirtualClusterStatNames& stat_names)
-        : stat_name_(stat_name), scope_(std::move(scope)),
+    VirtualClusterBase(absl::optional<std::string> vc_name, Stats::StatName stat_name,
+                       Stats::ScopePtr&& scope, const VirtualClusterStatNames& stat_names)
+        : vc_name_(vc_name), stat_name_(stat_name), scope_(std::move(scope)),
           stats_(generateStats(*scope_, stat_names)) {}
 
     // Router::VirtualCluster
+    const std::string& name() const override { return vc_name_; }
     Stats::StatName statName() const override { return stat_name_; }
     VirtualClusterStats& stats() const override { return stats_; }
 
   private:
+    const std::string vc_name_;
```

Comment:
rename to `name_` and add comment that `name_ `and `stat_name_` are two different representations for the same string, retained in memory to avoid symbol-table locks that would be needed when converting on-the-fly.